test/numa/run.sh

```diff
@@ -0,0 +1,482 @@
+        echo "Test input defaults:"
+        echo ""
+        echo "numanodes=${numanodes}"
+        echo ""
+        echo "cri_resmgr_cfg=${cri_resmgr_cfg}"
+        echo ""
+        echo -e "code=\"${code}\""
+        echo ""
+        echo "The defaults to QOSCLASS.yaml.in variables:"
+        echo "    ${yaml_in_defaults}"
+    elif [ "$2" == "script" ]; then
+        if [ "x$3" == "x" ]; then
+            help
+        else
+            help "$3"
+        fi
+    elif [ "x$2" == "x" ]; then
+        usage
+    else
+        echo "invalid help page, try:"
+        echo "  ./run.sh help"
+        echo "  ./run.sh help defaults"
+        echo "  ./run.sh help script [FUNCTION|all]"
+        exit 1
+    fi
+    exit 0
+elif [ "$mode" == "play" ]; then
+    speed=${speed-10}
+elif [ "$mode" == "test" ]; then
+    PV=
+elif [ "$mode" == "interactive" ]; then
+    PV=
+elif [ "$mode" == "record" ]; then
+    record
+else
+    usage
+    error "missing valid MODE"
+    exit 1
+fi
+
+if [ -n "$user_script_file" ]; then
+    if [ ! -f "$user_script_file" ]; then
+        error "cannot find test script file \"$user_script_file\""
+    fi
+    code=$(<"$user_script_file")
+fi
+
+# Prepare for test/demo
+mkdir -p "$OUTPUT_DIR"
+mkdir -p "$COMMAND_OUTPUT_DIR"
+rm -f "$COMMAND_OUTPUT_DIR"/0*
+( echo x > "$OUTPUT_DIR"/x && rm -f "$OUTPUT_DIR"/x ) || {
+    error "output directory outdir=$OUTPUT_DIR is not writable"
+}
+
+SUMMARY_FILE="$OUTPUT_DIR/summary.txt"
+echo -n "" > "$SUMMARY_FILE" || error "cannot write summary to \"$SUMMARY_FILE\""
+
+if [ "$binsrc" == "local" ]; then
+    [ -f "${BIN_DIR}/cri-resmgr" ] || error "missing \"${BIN_DIR}/cri-resmgr\""
+    [ -f "${BIN_DIR}/cri-resmgr-agent" ] || error "missing \"${BIN_DIR}/cri-resmgr-agent\""
+fi
+
+if [ -z "$VM_IP" ] || [ -z "$VM_SSH_USER" ] || [ -z "$VM_NAME" ]; then
+    screen-create-vm
+fi
+
+if ! vm-command-q "dpkg -l | grep -q kubelet"; then
+    screen-install-k8s
+fi
+
+if [ "$reinstall_cri_resmgr" == "1" ]; then
+    vm-command "kill -9 \$(pgrep cri-resmgr); rm -rf /usr/local/bin/cri-resmgr /usr/bin/cri-resmgr /usr/local/bin/cri-resmgr-agent /usr/bin/cri-resmgr-agent /var/lib/resmgr"
+fi
+
+if ! vm-command-q "[ -f /usr/local/bin/cri-resmgr ]"; then
+    screen-install-cri-resmgr
+fi
+
+# Start cri-resmgr if not already running
+if ! vm-command-q "pidof cri-resmgr" >/dev/null; then
+    screen-launch-cri-resmgr
+fi
+
+# Create kubernetes cluster or wait that it is online
+if vm-command-q "[ ! -f /var/lib/kubelet/config.yaml ]"; then
+    screen-create-singlenode-cluster
+else
+    # Wait for kube-apiserver to launch (may be down if the VM was just booted)
+    vm-wait-process kube-apiserver
+fi
+
+
+declare -A kind_count # associative arrays for counting created objects, like kind_count[pod]=1
+eval "${yaml_in_defaults}"
+if [ "$mode" == "interactive" ]; then
+    # shellcheck disable=SC2162
+    while read -e -p "run.sh> " -a commands; do
+        eval "${commands[@]}"
```

Comment:
I think this should be `eval "${commands[@]}" || :` instead.

Otherwise (now) a `vm-command` failure is just fine, while a 'built-in' command/helper failure (for instance, using `create` with a non-existent or mistyped YAML template name) ends the interactive session. I'm guessing this is unintentional. Since the lack of command-line history forces you to retype many of your commands all the time, this will sooner or later result in a built-in helper failing and the end of the session.

Reply:
I've felt the pain of user here as well. Been thrown out of the interactive mode when I forgot to protect myself from `exit` with parenthesis.

Part of me wanted to keep the pain point here, just in case someone someday would like to run
`cat script | ./run.sh interactive` (which doesn't even work at the moment... someone eats stdin before entering interactive mode). But I agree with you: let's make this nice in the user perspective. The name of the mode is `interactive` and not `stdin`, after all.

But the bad news is: `eval ... || :` will not solve the problem because evaluated script calls `exit 1` which causes the process to exit before evaluating the right-hand-side of `||`.

On the other hand, running `eval` always in subshell `( eval ... )` prevents using variables.

Making this nice to user may need a bigger change in the script runner design.

Reply:
Ok. As a band-aid, how about setting a variable for indicating interactive mode (INTERACTIVE_MODE=1) and doing `exit 1` in `command-error` only if that is not set ?

Reply:
Ended up patching this so that points where the same script command would have caused test script to exit either directly (`exit 1`) or indirectly (`command-error`, `verify False`), the interactive mode gets reset but it does not exit. This is definitely not perfect, but I didn't want to block only some of the errors from exiting the interactive mode...

Independently of the exit-on-error behavior, this change also makes `interactive` part of the script API, which I think may be quite useful. It's nice to let the test script execute somewhere where a bug might be close, and then enter the interactive mode for debugging.

Example: reproducing (with interactive pauses) a case where the number of reserved and allowed CPUs of a guaranteed pod drops from 4 to 3 after creating another guaranteed pod reserving one CPU. Yet  there would be a free NUMA node available without any reserved CPUs:

```
vm=c4m4x2 numanodes='[{"cpu": 4, "mem": "4G", "nodes": 2}]' code="CPU=4 create guaranteed; report cpus; interactive; create guaranteed; report cpus; interactive" ./run.sh test
```

Reply:
As a quick test I added this to command.bash
```
exit-if-not-interactive() {
    if [ -z "$INTERACTIVE_MODE" ]; then
        exit ${1:-1}
    fi
}
```
then replaced the most obvious offending `exit 1`'s in `command.bash:command-error()`, `run.sh:error()`, and `run.sh:verify()` with `exit-if-not-interactive 1`'s and set/cleared `INTERACTIVE_MODE=1` around the interactive mode `while`-loop.
It seemed to do the trick.

Reply:
Maybe we'll take this route. I'll integrate your approach and change my self-resetting-interactive mode. The thing I especially like here is that it allows interactive mode to change variable values when it is entered inside a test script.

Reply:
`exit-if-not-interactive` is now in.